File: android/src/main/java/com/snowplowanalytics/react/util/EventUtil.java
package com.snowplowanalytics.react.util;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableArray;
import com.snowplowanalytics.snowplow.event.DeepLinkReceived;
import com.snowplowanalytics.snowplow.event.MessageNotification;
import com.snowplowanalytics.snowplow.event.MessageNotificationAttachment;
import com.snowplowanalytics.snowplow.event.MessageNotificationTrigger;
import com.snowplowanalytics.snowplow.event.Structured;
import com.snowplowanalytics.snowplow.payload.SelfDescribingJson;
import com.snowplowanalytics.snowplow.event.ScreenView;
import com.snowplowanalytics.snowplow.event.PageView;
import com.snowplowanalytics.snowplow.event.Timing;
import com.snowplowanalytics.snowplow.event.ConsentGranted;
import com.snowplowanalytics.snowplow.event.ConsentWithdrawn;
import com.snowplowanalytics.snowplow.event.EcommerceTransactionItem;
import com.snowplowanalytics.snowplow.event.EcommerceTransaction;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

public class EventUtil {

    public static SelfDescribingJson createSelfDescribingJson(ReadableMap json) {
        String schema = json.getString("schema");
        ReadableMap dataMap = json.getMap("data");

        return new SelfDescribingJson(schema, dataMap.toHashMap());
    }

    public static List<SelfDescribingJson> createContexts(ReadableArray contexts) {
        List<SelfDescribingJson> nativeContexts = new ArrayList<>();
        for (int i = 0; i < contexts.size(); i++) {
            SelfDescribingJson json = createSelfDescribingJson(contexts.getMap(i));
            nativeContexts.add(json);
        }

        return nativeContexts;
    }

    public static Structured createStructuredEvent(ReadableMap argmap) {
        Structured event = new Structured(
                Objects.requireNonNull(argmap.getString("category"), "category can't be null"),
                Objects.requireNonNull(argmap.getString("action"), "action can't be null")
        );

        if (argmap.hasKey("label")) {
            event.label(argmap.getString("label"));
        }
        if (argmap.hasKey("property")) {
            event.property(argmap.getString("property"));
        }
        // React Native forces primitive double type - so null "value" parameter is handled by not setting at all
        if (argmap.hasKey("value") && !argmap.isNull("value")) {
            event.value(argmap.getDouble("value"));
        }

        return event;
    }

    public static ScreenView createScreenViewEvent(ReadableMap argmap) {
        @NonNull String name = Objects.requireNonNull(argmap.getString("name"), "name can't be null");
        ScreenView event = (
                argmap.hasKey("id") ?
                        new ScreenView(name, UUID.fromString(argmap.getString("id"))) :
                        new ScreenView(name)
        );

        if (argmap.hasKey("type")) {
            event.type(argmap.getString("type"));
        }
        if (argmap.hasKey("previousName")) {
            event.previousName(argmap.getString("previousName"));
        }
        if (argmap.hasKey("previousType")) {
            event.previousType(argmap.getString("previousType"));
        }
        if (argmap.hasKey("previousId")) {
            event.previousId(argmap.getString("previousId"));
        }
        if (argmap.hasKey("transitionType")) {
            event.transitionType(argmap.getString("transitionType"));
        }

        return event;
    }

    public static PageView createPageViewEvent(ReadableMap argmap) {
        PageView event = new PageView(
                Objects.requireNonNull(argmap.getString("pageUrl"), "pageUrl can't be null")
        );

        if (argmap.hasKey("pageTitle")) {
            event.pageTitle(argmap.getString("pageTitle"));
        }
        if (argmap.hasKey("referrer")) {
            event.referrer(argmap.getString("referrer"));
        }

        return event;
    }

    public static Timing createTimingEvent(ReadableMap argmap) {
        Timing event = new Timing(
                Objects.requireNonNull(argmap.getString("category"), "category can't be null"),
                Objects.requireNonNull(argmap.getString("variable"), "variable can't be null"),
                argmap.getInt("timing")
        );

        if (argmap.hasKey("label")) {
            event.label(argmap.getString("label"));
        }

        return event;
    }

    public static ConsentGranted createConsentGrantedEvent(ReadableMap argmap) {
        ConsentGranted event = new ConsentGranted(
                Objects.requireNonNull(argmap.getString("expiry"), "expiry can't be null"),
                Objects.requireNonNull(argmap.getString("documentId"), "documentId can't be null"),
                Objects.requireNonNull(argmap.getString("version"), "version can't be null")
        );

        if (argmap.hasKey("name")) {
            event.documentName(argmap.getString("name"));
        }
        if (argmap.hasKey("documentDescription")) {
            event.documentDescription(argmap.getString("documentDescription"));
        }

        return event;
    }

    public static ConsentWithdrawn createConsentWithdrawnEvent(ReadableMap argmap) {
        ConsentWithdrawn event = new ConsentWithdrawn(
                argmap.getBoolean("all"),
                Objects.requireNonNull(argmap.getString("documentId"), "documentId can't be null"),
                Objects.requireNonNull(argmap.getString("version"), "version can't be null")
        );

        if (argmap.hasKey("name")) {
            event.documentName(argmap.getString("name"));
        }
        if (argmap.hasKey("documentDescription")) {
            event.documentDescription(argmap.getString("documentDescription"));
        }

        return event;
    }

    public static List<EcommerceTransactionItem> createEcommerceTransactionItems(ReadableArray items) {
        List<EcommerceTransactionItem> ecomItems = new ArrayList<>();
        for (int i = 0; i < items.size(); i++) {
            ReadableMap argItem = items.getMap(i);
            EcommerceTransactionItem item = new EcommerceTransactionItem(
                    Objects.requireNonNull(argItem.getString("sku"), "sku can't be null"),
                    argItem.getDouble("price"),
                    argItem.getInt("quantity")
            );

            if (argItem.hasKey("name")) {
                item.name(argItem.getString("name"));
            }
            if (argItem.hasKey("category")) {
                item.category(argItem.getString("category"));
            }
            if (argItem.hasKey("currency")) {
                item.currency(argItem.getString("currency"));
            }

            ecomItems.add(item);
        }

        return ecomItems;
    }

    public static EcommerceTransaction createEcommerceTransactionEvent(ReadableMap argmap) {
        List<EcommerceTransactionItem> ecomItems = createEcommerceTransactionItems(
                Objects.requireNonNull(argmap.getArray("items"), "items can't be null")
        );

        EcommerceTransaction event = new EcommerceTransaction(
                Objects.requireNonNull(argmap.getString("orderId"), "orderId can't be null"),
                argmap.getDouble("totalValue"),
                ecomItems
        );

        if (argmap.hasKey("affiliation")) {
            event.affiliation(argmap.getString("affiliation"));
        }
        if (argmap.hasKey("taxValue")) {
            event.taxValue(argmap.getDouble("taxValue"));
        }
        if (argmap.hasKey("shipping")) {
            event.shipping(argmap.getDouble("shipping"));
        }
        if (argmap.hasKey("city")) {
            event.city(argmap.getString("city"));
        }
        if (argmap.hasKey("state")) {
            event.state(argmap.getString("state"));
        }
        if (argmap.hasKey("country")) {
            event.country(argmap.getString("country"));
        }
        if (argmap.hasKey("currency")) {
            event.currency(argmap.getString("currency"));
        }

        return event;
    }

    public static DeepLinkReceived createDeepLinkReceivedEvent(ReadableMap argmap) {
        DeepLinkReceived event = new DeepLinkReceived(
                Objects.requireNonNull(argmap.getString("url"), "url can't be null")
        );

        if (argmap.hasKey("referrer")) {
            event.referrer(argmap.getString("referrer"));
        }

        return event;
    }

    public static List<MessageNotificationAttachment> createMessageNotificationAttachments(ReadableArray items) {
        List<MessageNotificationAttachment> attachments = new ArrayList<>();
        for (int i = 0; i < items.size(); i++) {
            ReadableMap argItem = items.getMap(i);
            MessageNotificationAttachment attachment = new MessageNotificationAttachment(
                    Objects.requireNonNull(argItem.getString("identifier"), "identifier can't be null"),
                    Objects.requireNonNull(argItem.getString("type"), "type can't be null"),
                    Objects.requireNonNull(argItem.getString("url"), "url can't be null")
            );
            attachments.add(attachment);
        }

        return attachments;
    }

    public static List<String> createStrings(ReadableArray items) {
        List<String> results = new ArrayList<>();
        for (int i = 0; i < items.size(); i++) {
            results.add(items.getString(i));
        }
        return results;
    }

    public static MessageNotification createMessageNotificationEvent(ReadableMap argmap) {

        MessageNotificationTrigger trigger;
        switch (Objects.requireNonNull(argmap.getString("trigger"), "trigger can't be null")) {
            case "push":
                trigger = MessageNotificationTrigger.push;
                break;
            case "location":
                trigger = MessageNotificationTrigger.location;
                break;
            case "calendar":
                trigger = MessageNotificationTrigger.calendar;
                break;
            case "timeInterval":
                trigger = MessageNotificationTrigger.timeInterval;
                break;
            default:
                trigger = MessageNotificationTrigger.other;
                break;
        }

        MessageNotification event = new MessageNotification(
                Objects.requireNonNull(argmap.getString("title"), "title can't be null"),
                Objects.requireNonNull(argmap.getString("body"), "body can't be null"),
                trigger
        );

        if (argmap.hasKey("action")) {
            event.action(argmap.getString("action"));
        }
        if (argmap.hasKey("attachments")) {
            ReadableArray attachments = argmap.getArray("attachments");
            if (attachments != null) {
                event.attachments(createMessageNotificationAttachments(attachments));
            }
        }
        if (argmap.hasKey("bodyLocArgs")) {
            ReadableArray bodyLocArgs = argmap.getArray("bodyLocArgs");
            if (bodyLocArgs != null) {
                event.bodyLocArgs(createStrings(bodyLocArgs));
            }
        }
        if (argmap.hasKey("bodyLocKey")) {
            event.bodyLocKey(argmap.getString("bodyLocKey"));
        }
        if (argmap.hasKey("category")) {
            event.category(argmap.getString("category"));
        }
        if (argmap.hasKey("contentAvailable")) {
            event.contentAvailable(argmap.getBoolean("contentAvailable"));
        }
        if (argmap.hasKey("group")) {
            event.group(argmap.getString("group"));
        }
        if (argmap.hasKey("icon")) {
            event.icon(argmap.getString("icon"));
        }
        if (argmap.hasKey("notificationCount")) {
            event.notificationCount(argmap.getInt("notificationCount"));
        }
        if (argmap.hasKey("notificationTimestamp")) {
            event.notificationTimestamp(argmap.getString("notificationTimestamp"));
        }
        if (argmap.hasKey("sound")) {
            event.sound(argmap.getString("sound"));
        }
        if (argmap.hasKey("subtitle")) {
            event.subtitle(argmap.getString("subtitle"));
        }
        if (argmap.hasKey("tag")) {
            event.tag(argmap.getString("tag"));
        }
        if (argmap.hasKey("threadIdentifier")) {
            event.threadIdentifier(argmap.getString("threadIdentifier"));
        }
        if (argmap.hasKey("titleLocArgs")) {
            ReadableArray titleLocArgs = argmap.getArray("titleLocArgs");
            if (titleLocArgs != null) {
                event.titleLocArgs(createStrings(titleLocArgs));
            }
        }
        if (argmap.hasKey("titleLocKey")) {
            event.titleLocKey(argmap.getString("titleLocKey"));
        }

        return event;
    }
}

File: android/src/main/java/com/snowplow/reactnativetracker/util/EventUtil.kt
package com.snowplow.reactnativetracker.util

import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.snowplowanalytics.snowplow.event.*
import com.snowplowanalytics.snowplow.payload.SelfDescribingJson
import java.util.*

object EventUtil {
  fun createSelfDescribingJson(json: ReadableMap): SelfDescribingJson {
    val schema: String = json.getString("schema")!!
    val dataMap: ReadableMap = json.getMap("data")!!
    return SelfDescribingJson(schema, dataMap.toHashMap())
  }

  fun createContexts(contexts: ReadableArray): List<SelfDescribingJson> {
    val nativeContexts: MutableList<SelfDescribingJson> = ArrayList()
    for (i in 0 until contexts.size()) {
      val json = createSelfDescribingJson(contexts.getMap(i))
      nativeContexts.add(json)
    }
    return nativeContexts
  }

  fun createStructuredEvent(argmap: ReadableMap): Structured {
    val event = Structured(
      argmap.getString("category")!!,
      argmap.getString("action")!!
    )
    argmap.getString("label")?.let { event.label(it) }
    argmap.getString("property")?.let { event.property(it) }
    // React Native forces primitive double type - so null "value" parameter is handled by not setting at all
    if (argmap.hasKey("value") && !argmap.isNull("value")) {
      event.value(argmap.getDouble("value"))
    }
    return event
  }

  fun createScreenViewEvent(argmap: ReadableMap): ScreenView {
    val name: String = argmap.getString("name")!!
    val event = if (argmap.hasKey("id")) ScreenView(
      name,
      UUID.fromString(argmap.getString("id"))
    ) else ScreenView(name)
    argmap.getString("type")?.let { event.type(it) }
    argmap.getString("previousName")?.let { event.previousName(it) }
    argmap.getString("previousType")?.let { event.previousType(it) }
    argmap.getString("previousId")?.let { event.previousId(it) }
    argmap.getString("transitionType")?.let { event.transitionType(it) }
    return event
  }

  fun createScrollChangedEvent(argmap: ReadableMap): ScrollChanged {
    val event = ScrollChanged(
      yOffset = if (argmap.hasKey("yOffset")) { argmap.getDouble("yOffset").toInt() } else { null },
      xOffset = if (argmap.hasKey("xOffset")) { argmap.getDouble("xOffset").toInt() } else { null },
      viewHeight = if (argmap.hasKey("viewHeight")) { argmap.getDouble("viewHeight").toInt() } else { null },
      viewWidth = if (argmap.hasKey("viewWidth")) { argmap.getDouble("viewWidth").toInt() } else { null },
      contentHeight = if (argmap.hasKey("contentHeight")) { argmap.getDouble("contentHeight").toInt() } else { null },
      contentWidth = if (argmap.hasKey("contentWidth")) { argmap.getDouble("contentWidth").toInt() } else { null },
    )
    return event
  }

  fun createListItemViewEvent(argmap: ReadableMap): ListItemView {
    val event = ListItemView(
      index = argmap.getDouble("index").toInt(),
      itemsCount = if (argmap.hasKey("itemsCount")) { argmap.getDouble("itemsCount").toInt() } else { null },
    )
    return event
  }

  fun createPageViewEvent(argmap: ReadableMap): PageView {
    val event = PageView(argmap.getString("pageUrl")!!)
    argmap.getString("pageTitle")?.let { event.pageTitle(it) }
    argmap.getString("referrer")?.let { event.referrer(it) }
    return event
  }

  fun createTimingEvent(argmap: ReadableMap): Timing {
    val event = Timing(
      argmap.getString("category")!!,
      argmap.getString("variable")!!,
      argmap.getDouble("timing").toInt()
    )
    argmap.getString("label")?.let { label ->
      event.label(label)
    }
    return event
  }

  fun createConsentGrantedEvent(argmap: ReadableMap): ConsentGranted {
    val event = ConsentGranted(
      argmap.getString("expiry")!!,
      argmap.getString("documentId")!!,
      argmap.getString("version")!!
    )
    argmap.getString("name")?.let { event.documentName(it) }
    argmap.getString("documentDescription")?.let { event.documentDescription(it) }
    return event
  }

  fun createConsentWithdrawnEvent(argmap: ReadableMap): ConsentWithdrawn {
    val event = ConsentWithdrawn(
      argmap.getBoolean("all"),
      argmap.getString("documentId")!!,
      argmap.getString("version")!!
    )
    argmap.getString("name")?.let { event.documentName(it) }
    argmap.getString("documentDescription")?.let { event.documentDescription(it) }
    return event
  }

  fun createEcommerceTransactionItems(items: ReadableArray): List<EcommerceTransactionItem> {
    val ecomItems: MutableList<EcommerceTransactionItem> = ArrayList()
    for (i in 0 until items.size()) {
      val argItem: ReadableMap = items.getMap(i)
      val item = EcommerceTransactionItem(
        argItem.getString("sku")!!,
        argItem.getDouble("price"),
        argItem.getDouble("quantity").toInt()
      )
      argItem.getString("name")?.let { item.name(it) }
      argItem.getString("category")?.let { item.category(it) }
      argItem.getString("currency")?.let { item.currency(it) }
      ecomItems.add(item)
    }
    return ecomItems
  }

  fun createEcommerceTransactionEvent(argmap: ReadableMap): EcommerceTransaction {
    val ecomItems = createEcommerceTransactionItems(
      argmap.getArray("items")!!
    )
    val event = EcommerceTransaction(
      argmap.getString("orderId")!!,
      argmap.getDouble("totalValue"),
      ecomItems
    )
    argmap.getString("affiliation")?.let { event.affiliation(it) }
    if (argmap.hasKey("taxValue")) {
      event.taxValue(argmap.getDouble("taxValue"))
    }
    if (argmap.hasKey("shipping")) {
      event.shipping(argmap.getDouble("shipping"))
    }
    argmap.getString("city")?.let { event.city(it) }
    argmap.getString("state")?.let { event.state(it) }
    argmap.getString("country")?.let { event.country(it) }
    argmap.getString("currency")?.let { event.currency(it) }
    return event
  }

  fun createDeepLinkReceivedEvent(argmap: ReadableMap): DeepLinkReceived {
    val event = DeepLinkReceived(
      argmap.getString("url")!!
    )
    argmap.getString("referrer")?.let { event.referrer(it) }
    return event
  }

  fun createMessageNotificationAttachments(items: ReadableArray): List<MessageNotificationAttachment> {
    val attachments: MutableList<MessageNotificationAttachment> = ArrayList()
    for (i in 0 until items.size()) {
      val argItem: ReadableMap = items.getMap(i)
      val attachment = MessageNotificationAttachment(
        argItem.getString("identifier")!!,
        argItem.getString("type")!!,
        argItem.getString("url")!!
      )
      attachments.add(attachment)
    }
    return attachments
  }

  fun createStrings(items: ReadableArray): List<String> {
    val results: MutableList<String> = ArrayList()
    for (i in 0 until items.size()) {
      results.add(items.getString(i))
    }
    return results
  }

  fun createMessageNotificationEvent(argmap: ReadableMap): MessageNotification {
    val trigger: MessageNotificationTrigger
    trigger =
      when (argmap.getString("trigger")!!) {
        "push" -> MessageNotificationTrigger.push
        "location" -> MessageNotificationTrigger.location
        "calendar" -> MessageNotificationTrigger.calendar
        "timeInterval" -> MessageNotificationTrigger.timeInterval
        else -> MessageNotificationTrigger.other
      }
    val event = MessageNotification(
      argmap.getString("title")!!,
      argmap.getString("body")!!,
      trigger
    )
    argmap.getString("action")?.let { event.action(it) }
    if (argmap.hasKey("attachments")) {
      argmap.getArray("attachments")?.let { attachments ->
        event.attachments(createMessageNotificationAttachments(attachments))
      }
    }
    if (argmap.hasKey("bodyLocArgs")) {
      argmap.getArray("bodyLocArgs")?.let { bodyLocArgs ->
        event.bodyLocArgs(createStrings(bodyLocArgs))
      }
    }
    argmap.getString("bodyLocKey")?.let { event.bodyLocKey(it) }
    argmap.getString("category")?.let { event.category(it) }
    if (argmap.hasKey("contentAvailable")) {
      event.contentAvailable(argmap.getBoolean("contentAvailable"))
    }
    argmap.getString("group")?.let { event.group(it) }
    argmap.getString("icon")?.let { event.icon(it) }
    if (argmap.hasKey("notificationCount")) {
      event.notificationCount(argmap.getDouble("notificationCount").toInt())
    }
    argmap.getString("notificationTimestamp")?.let { event.notificationTimestamp(it) }
    argmap.getString("sound")?.let { event.sound(it) }
    argmap.getString("subtitle")?.let { event.subtitle(it) }
    argmap.getString("tag")?.let { event.tag(it) }
    argmap.getString("threadIdentifier")?.let { event.threadIdentifier(it) }
    if (argmap.hasKey("titleLocArgs")) {
      argmap.getArray("titleLocArgs")?.let { titleLocArgs ->
        event.titleLocArgs(createStrings(titleLocArgs))
      }
    }
    argmap.getString("titleLocKey")?.let { event.titleLocKey(it) }
    return event
  }
}
